File: frontend/src/lib/store/sandbox.ts
'use client';

import { create } from 'zustand';

export type SandboxSessionStatus = 'draft' | 'running' | 'paused';
export type SandboxActorClass =
  | 'orbital'
  | 'fixed_ground'
  | 'mobile_ground'
  | 'air'
  | 'sea'
  | 'weapon'
  | 'effect';
export type SandboxFaction = 'allied' | 'hostile' | 'neutral' | 'unknown';
export type SandboxProvenance = 'manual' | 'agent' | 'live_cloned';
export type SandboxScenarioItemType = 'event' | 'modifier' | 'overlay' | 'objective';

export type SandboxInteractionMode =
  | 'idle'
  | 'place_template'
  | 'relocate_actor'
  | 'set_move_target'
  | 'add_waypoint'
  | 'place_marker'
  | 'draw_route'
  | 'draw_area';

export type SandboxContextTab = 'actors' | 'import' | 'saved' | 'ground';

export interface SandboxSession {
  id: string;
  tenant_id: string;
  user_id: string;
  name: string;
  status: SandboxSessionStatus;
  is_saved: boolean;
  description?: string | null;
  initial_prompt?: string | null;
  current_time_seconds: number;
  time_multiplier: number;
  duration_seconds: number | null;
  created_at: string;
  updated_at: string;
}

export interface SandboxPosition {
  lat: number;
  lon: number;
  alt_m: number;
}

export interface SandboxActor {
  id: string;
  session_id: string;
  tenant_id: string;
  actor_class: SandboxActorClass;
  actor_type: string;
  subtype?: string | null;
  faction: SandboxFaction;
  label: string;
  provenance: SandboxProvenance;
  visual_config: Record<string, unknown>;
  state: Record<string, unknown>;
  initial_state: Record<string, unknown>;
  capabilities: Record<string, unknown>;
  behavior: Record<string, unknown>;
  source_ref: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SandboxScenarioItem {
  id: string;
  session_id: string;
  tenant_id: string;
  item_type: SandboxScenarioItemType;
  label: string;
  source_type?: string | null;
  source_id?: string | null;
  payload: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SandboxCommand {
  id: string;
  session_id: string;
  tenant_id: string;
  command_type: string;
  source: 'manual' | 'chat' | 'import' | 'system';
  summary: string;
  payload: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SandboxSnapshot {
  session: SandboxSession;
  actors: SandboxActor[];
  scenario_items: SandboxScenarioItem[];
  commands: SandboxCommand[];
}

// --------------- GROUND PLANNING TYPES ---------------

export type TacticalMarkerType = 'objective' | 'rally_point' | 'op' | 'hq' | 'checkpoint';
export type TacticalRouteType = 'attack_axis' | 'retreat_route' | 'patrol_route' | 'supply_route' | 'phase_line';
export type TacticalAreaType = 'ao' | 'kill_zone' | 'safe_zone' | 'restricted' | 'objective_area';

export interface TacticalMarker {
  id: string;
  markerType: TacticalMarkerType;
  label: string;
  position: SandboxPosition;
  faction: SandboxFaction;
}

export interface TacticalRoute {
  id: string;
  routeType: TacticalRouteType;
  label: string;
  points: SandboxPosition[];
  faction: SandboxFaction;
}

export interface TacticalArea {
  id: string;
  areaType: TacticalAreaType;
  label: string;
  vertices: SandboxPosition[];
  faction: SandboxFaction;
}

export interface GroundPlan {
  markers: TacticalMarker[];
  routes: TacticalRoute[];
  areas: TacticalArea[];
}

export interface GroundDrawingConfig {
  markerType: TacticalMarkerType;
  routeType: TacticalRouteType;
  areaType: TacticalAreaType;
  label: string;
  faction: SandboxFaction;
}

export interface SandboxTemplateDraft {
  actorClass: SandboxActorClass;
  actorType: string;
  subtype?: string | null;
  label: string;
  faction: SandboxFaction;
  behavior?: Record<string, unknown>;
  capabilities?: Record<string, unknown>;
  visualConfig?: Record<string, unknown>;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
}

interface SandboxState {
  snapshot: SandboxSnapshot | null;
  selectedActorId: string | null;
  interactionMode: SandboxInteractionMode;
  interactionPayload: SandboxTemplateDraft | null;
  isBootstrapping: boolean;
  contextPanelOpen: boolean;
  contextTab: SandboxContextTab;
  chatMessages: ChatMessage[];
  chatBusy: boolean;
  groundPlan: GroundPlan;
  drawingPoints: SandboxPosition[];
  groundDrawingConfig: GroundDrawingConfig;

  setBootstrapping: (value: boolean) => void;
  hydrateSnapshot: (snapshot: SandboxSnapshot) => void;
  selectActor: (actorId: string | null) => void;
  setInteractionMode: (mode: SandboxInteractionMode, payload?: SandboxTemplateDraft | null) => void;
  setContextPanelOpen: (open: boolean) => void;
  setContextTab: (tab: SandboxContextTab) => void;
  appendChat: (message: ChatMessage) => void;
  setChatBusy: (busy: boolean) => void;
  addTacticalMarker: (marker: TacticalMarker) => void;
  removeTacticalElement: (elementType: 'marker' | 'route' | 'area', id: string) => void;
  addTacticalRoute: (route: TacticalRoute) => void;
  addTacticalArea: (area: TacticalArea) => void;
  addDrawingPoint: (point: SandboxPosition) => void;
  clearDrawing: () => void;
  setGroundDrawingConfig: (config: Partial<GroundDrawingConfig>) => void;
  clearGroundPlan: () => void;
  reset: () => void;
}

export function buildChatMessage(role: ChatMessage['role'], content: string): ChatMessage {
  return {
    id: `${role}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    role,
    content,
  };
}

export const useSandboxStore = create<SandboxState>((set) => ({
  snapshot: null,
  selectedActorId: null,
  interactionMode: 'idle',
  interactionPayload: null,
  isBootstrapping: true,
  contextPanelOpen: true,
  contextTab: 'actors',
  chatMessages: [],
  chatBusy: false,
  groundPlan: { markers: [], routes: [], areas: [] },
  drawingPoints: [],
  groundDrawingConfig: { markerType: 'objective', routeType: 'attack_axis', areaType: 'ao', label: '', faction: 'allied' },

  setBootstrapping: (value) => set({ isBootstrapping: value }),
  hydrateSnapshot: (snapshot) =>
    set((state) => ({
      snapshot,
      selectedActorId:
        state.selectedActorId && snapshot.actors.some((a) => a.id === state.selectedActorId)
          ? state.selectedActorId
          : null,
    })),
  selectActor: (actorId) => set({ selectedActorId: actorId }),
  setInteractionMode: (mode, payload) =>
    set({ interactionMode: mode, interactionPayload: payload ?? null }),
  setContextPanelOpen: (open) => set({ contextPanelOpen: open }),
  setContextTab: (tab) => set({ contextTab: tab }),
  appendChat: (message) =>
    set((state) => ({ chatMessages: [...state.chatMessages, message] })),
  setChatBusy: (busy) => set({ chatBusy: busy }),
  addTacticalMarker: (marker) =>
    set((s) => ({ groundPlan: { ...s.groundPlan, markers: [...s.groundPlan.markers, marker] } })),
  removeTacticalElement: (elementType, id) =>
    set((s) => ({
      groundPlan: {
        markers: elementType === 'marker' ? s.groundPlan.markers.filter((m) => m.id !== id) : s.groundPlan.markers,
        routes: elementType === 'route' ? s.groundPlan.routes.filter((r) => r.id !== id) : s.groundPlan.routes,
        areas: elementType === 'area' ? s.groundPlan.areas.filter((a) => a.id !== id) : s.groundPlan.areas,
      },
    })),
  addTacticalRoute: (route) =>
    set((s) => ({ groundPlan: { ...s.groundPlan, routes: [...s.groundPlan.routes, route] } })),
  addTacticalArea: (area) =>
    set((s) => ({ groundPlan: { ...s.groundPlan, areas: [...s.groundPlan.areas, area] } })),
  addDrawingPoint: (point) =>
    set((s) => ({ drawingPoints: [...s.drawingPoints, point] })),
  clearDrawing: () => set({ drawingPoints: [] }),
  setGroundDrawingConfig: (config) =>
    set((s) => ({ groundDrawingConfig: { ...s.groundDrawingConfig, ...config } })),
  clearGroundPlan: () => set({ groundPlan: { markers: [], routes: [], areas: [] }, drawingPoints: [] }),
  reset: () =>
    set({
      snapshot: null,
      selectedActorId: null,
      interactionMode: 'idle',
      interactionPayload: null,
      isBootstrapping: true,
      contextPanelOpen: true,
      contextTab: 'actors',
      chatMessages: [],
      chatBusy: false,
      groundPlan: { markers: [], routes: [], areas: [] },
      drawingPoints: [],
      groundDrawingConfig: { markerType: 'objective', routeType: 'attack_axis', areaType: 'ao', label: '', faction: 'allied' },
    }),
}));

File: backend/app/db/models/sandbox.py
"""Sandbox models for isolated scenario authoring and runtime state."""
from sqlalchemy import Boolean, Column, Float, ForeignKey, Index, JSON, String, Text
from sqlalchemy.orm import relationship

from app.db.base import Base, AuditMixin, generate_uuid


class SandboxSession(Base, AuditMixin):
    """Per-user isolated sandbox workspace."""

    __tablename__ = "sandbox_sessions"

    id = Column(String(50), primary_key=True, default=generate_uuid)
    user_id = Column(String(100), nullable=False, index=True)
    name = Column(String(200), nullable=False, default="Untitled Sandbox")
    description = Column(Text, nullable=True)
    status = Column(String(20), nullable=False, default="draft", index=True)
    is_saved = Column(Boolean, nullable=False, default=False)
    initial_prompt = Column(Text, nullable=True)
    current_time_seconds = Column(Float, nullable=False, default=0.0)
    time_multiplier = Column(Float, nullable=False, default=1.0)
    duration_seconds = Column(Float, nullable=True)

    actors = relationship(
        "SandboxActor",
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="SandboxActor.created_at",
    )
    scenario_items = relationship(
        "SandboxScenarioItem",
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="SandboxScenarioItem.created_at",
    )
    commands = relationship(
        "SandboxCommand",
        back_populates="session",
        cascade="all, delete-orphan",
        order_by="SandboxCommand.created_at",
    )

    __table_args__ = (
        Index("ix_sandbox_sessions_tenant_user", "tenant_id", "user_id"),
    )


class SandboxActor(Base, AuditMixin):
    """Sandbox-local actor cloned from live data or created manually."""

    __tablename__ = "sandbox_actors"

    id = Column(String(50), primary_key=True, default=generate_uuid)
    session_id = Column(
        String(50),
        ForeignKey("sandbox_sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    actor_class = Column(String(40), nullable=False, index=True)
    actor_type = Column(String(40), nullable=False, index=True)
    subtype = Column(String(80), nullable=True)
    faction = Column(String(20), nullable=False, default="neutral")
    label = Column(String(200), nullable=False)
    provenance = Column(String(30), nullable=False, default="manual")
    visual_config = Column(JSON, nullable=False, default=dict)
    state = Column(JSON, nullable=False, default=dict)
    initial_state = Column(JSON, nullable=False, default=dict)
    capabilities = Column(JSON, nullable=False, default=dict)
    behavior = Column(JSON, nullable=False, default=dict)
    source_ref = Column(JSON, nullable=False, default=dict)

    session = relationship("SandboxSession", back_populates="actors")

    __table_args__ = (
        Index("ix_sandbox_actors_session_label", "session_id", "label"),
        Index("ix_sandbox_actors_tenant_type", "tenant_id", "actor_type"),
    )


class SandboxScenarioItem(Base, AuditMixin):
    """Sandbox-local non-actor imports like events, modifiers, and objectives."""

    __tablename__ = "sandbox_scenario_items"

    id = Column(String(50), primary_key=True, default=generate_uuid)
    session_id = Column(
        String(50),
        ForeignKey("sandbox_sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    item_type = Column(String(30), nullable=False, index=True)
    label = Column(String(200), nullable=False)
    source_type = Column(String(50), nullable=True)
    source_id = Column(String(100), nullable=True)
    payload = Column(JSON, nullable=False, default=dict)

    session = relationship("SandboxSession", back_populates="scenario_items")

    __table_args__ = (
        Index("ix_sandbox_items_session_type", "session_id", "item_type"),
    )


class SandboxCommand(Base, AuditMixin):
    """Replayable sandbox command history."""

    __tablename__ = "sandbox_commands"

    id = Column(String(50), primary_key=True, default=generate_uuid)
    session_id = Column(
        String(50),
        ForeignKey("sandbox_sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    command_type = Column(String(60), nullable=False, index=True)
    source = Column(String(20), nullable=False, default="manual")
    summary = Column(Text, nullable=False)
    payload = Column(JSON, nullable=False, default=dict)

    session = relationship("SandboxSession", back_populates="commands")

    __table_args__ = (
        Index("ix_sandbox_commands_session_created", "session_id", "created_at"),
    )
